tracer/src/Datadog.Tracer.Native/debugger_rejit_preprocessor.cpp

```diff
@@ -250,9 +253,166 @@ void DebuggerRejitPreprocessor::UpdateMethod(RejitHandlerModuleMethod* methodHan
+            if (FAILED(hr))
+            {
+                Logger::Error("DebuggerRejitPreprocessor::GetMoveNextMethodFromKickOffMethod: failed to call metadataImport->FindMethod for MoveNext method");
+                return  hr;
+            }
+        }
+
+        if (moveNextMethod != mdMethodDefNil)
+        {
+            // we found the correct nested type and MoveNext method
+            break;
+        }
+    }
+    return S_OK;
+}
+
+std::tuple<HRESULT, mdMethodDef, FunctionInfo> DebuggerRejitPreprocessor::TransformKickOffToMoveNext(
+                                                    const ComPtr<IMetaDataImport2>& metadataImport, const ComPtr<IMetaDataEmit2>& metadataEmit,
+                                                    mdMethodDef moveNextMethod, mdTypeDef nestedAsyncClassOrStruct)
+{
+    // this is an async method and we found the generated nested state machine type,
+    // define a new boolean field in the state machine object to indicate whether we have already entered the MoveNext method
+    // (if we have, it means we are re-entering the method as a continuation in a subsequent `await` operation, 
+    // and should not capture the method parameter values as we do the first time around).
+    constexpr BYTE fieldSignature[] = {IMAGE_CEE_CS_CALLCONV_FIELD, ELEMENT_TYPE_BOOLEAN};
+    mdFieldDef isFirstEntry = mdFieldDefNil;
+    auto hr = metadataEmit->DefineField(nestedAsyncClassOrStruct, managed_profiler_debugger_is_first_entry_field_name.c_str(), fdPrivate, fieldSignature,
+                                   sizeof(fieldSignature), 0, nullptr, 0, &isFirstEntry);
+    if (FAILED(hr))
+    {
+        Logger::Error("DebuggerRejitPreprocessor::TransformKickOffToMoveNext: DefineField _isFirstEntry failed");
+        return {hr, mdMethodDefNil, FunctionInfo()};
+    }
+
+    // save the MoveNext method and create a function info for it
+    auto caller = GetFunctionInfo(metadataImport, moveNextMethod);
+    if (!caller.IsValid())
+    {
+        Logger::Error("DebuggerRejitPreprocessor::TransformKickOffToMoveNext: The caller for the methoddef: ",
+                      shared::TokenStr(&moveNextMethod), " is not valid!");
+        return {E_FAIL, mdMethodDefNil, FunctionInfo()};
+    }
+
+    hr = caller.method_signature.TryParse();
+    if (FAILED(hr))
+    {
+        Logger::Error("DebuggerRejitPreprocessor::TransformKickOffToMoveNext: The method signature: ",
+                      caller.method_signature.str(), " cannot be parsed.");
+        return {hr, mdMethodDefNil, FunctionInfo()};
+    }
+
+    return {S_OK, moveNextMethod, FunctionInfo(caller)};
+}
+
+/**
+ * \brief In certain cases, the method that we have in our hand (methodDef, typeDef and functionInfo) is not what we need to instrument,
+ * e.g. in async method we have the kickoff method but we need to instrument the MoveNext method,
+ * we may be required also to do some manipulation to peek the correct method in the case of hierarchy
+ */
```

Comment:
```suggestion
 * e.g. in async method we have the kickoff method but we need to instrument the MoveNext method.
 * In the future, this may be applicable to other cases (e.g. if the user is trying to instrument an abstract method, we should actually instrument the methods that override and implement it, etc etc).
 */
```

Reply:
Fixed in fe1d802401001b1262ffa124f56b37f25666f52e